### DISCUSSION TITLE
Expose the class and call new() on it

### DISCUSSION THREAD

**Discussion by 0xbuidler** [paulrberg/hardhat-packager#7] [General]
At first, I thought that the idea of the packager was to expose classes from the typechain to a npm package way in order to import the typechain anywhere (another hh project maybe) and just call new on a factory right away. But this plugin doesn't (yet ?). Any plans on that in the future ? Cause I see that you are exporting abi, bytecode and then recreating the factory in your exemples. I've tried myself to publish a hardhat project to npm and exposing as well the typechain directly but it doesn't seems to work when importing from another hardhat project. All i'm getting is either `SyntaxError: Unexpected token 'export'` or `SyntaxError: Cannot use import statement outside a module`. It's a common error, I know, but I cannot make it work

**Comment by PaulRBerg** [comment]
Are you using [v1.1.0](https://github.com/paulrberg/hardhat-packager/releases/tag/v1.1.0)? Support for factories has been added in that release. Just make sure that the `includeFactories` flag is set to true in your `packager` config.

Example for how to use the factories in a Hardhat task, assuming that `your-package` is your package published on npm:

```ts
import { YourContract__factory} from "your-package/typechain/factories/YourContract__factory";
import { task } from "hardhat/config";

task("deploy:YourContract").setAction(async function (_, { ethers }) {
  const signers: SignerWithAddress[] = await ethers.getSigners();
  const chainlinkOperatorFactory: YourContract__factory = new YourContract__factory(signers[0]);
  const yourContract: YourContract = await yourContractFactory.deploy();
  await yourContract.deployed();
  console.log("YourContract deployed to: ", yourContract.address);
});

**Reply by 0xbuidler** [reply]
I doubt this will work. If the contract is from another package, hardhat won't have it stored hence `ethers.getContractFactory("")` will fail to find the contract.

On another hand, I tried again with your plugin and same thing happened ... `SyntaxError: Cannot use import statement outside a module`

**Reply by PaulRBerg** [reply]
Oh shoot, you're right. I updated the script to use the constructor of the factory, and pass in the first signed returned by ethers.

To be frank I kinda rushed with the factories feature; I added support for them because a user told me privately that he was impatient. This is what happens when I don't do proper integration testing, and give too much credence to unit testing.

That said, I think that there might be a way out, such that we keep the factories feature while making them usable with external Hardhat projects, like you want to. I opened an issue for the error you reported: https://github.com/paulrberg/hardhat-packager/issues/8. After that is fixed, and I ship a new version of the plugin, you should be able to use it and instantiate factories.

**Reply by 0xbuidler** [reply]
Thanks ! :)

**Comment by 0xbuidler** [comment]
Ok so I found a solution: 
1 - `yarn build`
2 - set the tsconfig.json to only include the `./typechain` folder
3 - set into `package.json`:
```
  "main": "typechain/index.js",
  "types": "typechain/index.d.ts",
  "files": [
    "/contracts/**/*.sol",
    "/artifacts/contracts/**/*.json",
    "/typechain"
  ],
```

4 - `tsc`
5 - npm publish your package

**Reply by PaulRBerg** [reply]
This works, but I have two notes:

1. You may like to add a glob that ignores some folders in the `contracts` folder. E.g. `test`. I do this by removing the file extension from the glob, and adding an `.npmignore` file in the `contracts` folder.
2. You may have to remove the `/contracts` path from the artifacts glob. `hardhat-packager` doesn't add such folder.

**Comment by PaulRBerg** [comment]
I was scouring the web for a way to publish TypeScript code to the npm registry, and make it possible to import it directly from there. But it looks like only JavaScript is supported for now.

I will prevent the TypeChain factories from being included by default, and I will update the README to explain how to handle them. This is not such a straightforward a matter as I had initially thought. Once again, thanks @sshmaxime for your updates.

**Reply by 0xbuidler** [reply]
No worries, glad I could help a bit !
